### PR TITLE
logs: make sure log-row-react-keys are always unique

### PR DESCRIPTION
**Pull request by gabor**
we use the `LogRowModel.uid` attribute, that's supposed to be unique, as react-keys in log-rows.

the problem is, with logs-context-infinite-scroll, we keep adding new rows to the list of existing rows, and duplication may happen. (we keep fixing these issues, but it would be good to be robust against this problem).

this PR makes sure that every log-row has an unique index. it tries first to use the provided `.uid`, but if there's a duplicate for it, it will go for an index-based value.

how to test:
1. (currently the elastic datasource has a problem that generates duplicate uids for infinite-scroll, so it's a good test case)
2. run elastic with some data, open logs-context
3. scroll to trigger the "load more" functionality
4. at this point on main-branch, you will see in the console a large amount of react-duplicate-keys errors. with this PR, there will be no errors.

### CODE REVIEW

File: public/app/features/logs/components/LogRows.tsx [hunk 1]
```diff
@@ -56,6 +56,29 @@ interface State {
   renderAll: boolean;
 }
 
+// this class generates react-keys that are guaranteed to be unique.
```
Comment (matyax) [inline public/app/features/logs/components/LogRows.tsx:59]: I would 100% move this to its own module. Not only for organization, but to allow it to be used in other places.

File: public/app/features/logs/components/LogRows.tsx [hunk 2]
```diff
@@ -56,6 +56,29 @@ interface State {
...
+// it will use an index-based key.
+class KeyMaker {
+  seen: Set<string>;
+  count: number;
+  constructor() {
+    this.seen = new Set();
+    this.count = 0;
+  }
+
+  getKey(maybeId: string) {
```
Comment (matyax) [inline public/app/features/logs/components/LogRows.tsx:70]: Either

```javascript
  getKey(maybeId: string) {
```

or

```javascript
class UniqueKeyMaker {
```

**Review by matyax — COMMENTED** [review]
Great solution! Would you mind adding a few test cases for `KeyMaker`.

File: public/app/features/logs/components/LogRows.tsx (hunk 2 above)
Comment (gabor) [reply public/app/features/logs/components/LogRows.tsx:70]: sorry, could you explain this in more detail? thanks!

Comment (matyax) [reply public/app/features/logs/components/LogRows.tsx:70]: Sure! Just edited my comment, I forgot to change the function name.

TL;DR it's to get unique keys so either add unique to the class or to the method.

Comment (gabor) [reply public/app/features/logs/components/LogRows.tsx]: i see. thanks! updated.

File: public/app/features/logs/components/LogRows.tsx (hunk 1 above)
Comment (gabor) [reply public/app/features/logs/components/LogRows.tsx]: sure! moved it to a separate file in the logs "area".

**Comment by gabor** [comment]
> Great solution! Would you mind adding a few test cases for `KeyMaker`.

good point. added!

File: public/app/features/logs/UniqueKeyMaker.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  getKey(proposedKey: string) {
+    this.count += 1;
+    const maybeKey = `k_${proposedKey}`;
+    if (this.seen.has(maybeKey)) {
+      return `i_${this.count}`;
+    } else {
+      this.seen.add(maybeKey);
+      return maybeKey;
+    }
+  }
```
Comment (matyax) [inline public/app/features/logs/UniqueKeyMaker.ts:24]: At this point this feels nitpicky, but maybe you'd agree. I would only override the key if it's seen, otherwise I'll return it unmodified.

```suggestion
  getKey(proposedKey: string) {
    this.count += 1;
    if (this.seen.has(proposedKey)) {
      // I would also consider 1) keeping the original key and just prefixing it and 2) pass the new key through the `seen` check to be 100% sure it's not duplicated.
      return this.getKey(`i_${this.count}_${proposedKey}`);
    } else {
      this.seen.add(proposedKey);
      return proposedKey;
    }
  }
```

**Review by matyax — APPROVED** [review]
Other than one nitpicky comment, great work!

File: public/app/features/logs/UniqueKeyMaker.ts (hunk 3 above)
Comment (gabor) [reply public/app/features/logs/UniqueKeyMaker.ts:24]: i see your point, but the problem is that that does not guarantee 100% uniqueness..
imagine these keys:
- one
- i_3
- one

currently this produces the following keys:
- k_one
- k_i_3
- i_3


if we don't prefix unique keys, it will produce this:
- one
- i_3
- i_3

which is bad.

technically we could also `seen`-check the index-based values, but then there's a problem. when you seen-check an indexed-value, and see that it has been used, then what? increment the index and try again? what if that one is used too? again increment? how many times? etc.

i don't think it's unsolvable, but right now i don't have any ideas that would be simple and do a 100% guarantee... but i'm very open to ideas.  i will merge this one, i hope that's ok, and we can improve it if we come up with a good idea 👍

Comment (matyax) [reply public/app/features/logs/UniqueKeyMaker.ts:24]: Makes sense! Thanks for checking.
